### PR TITLE
postcard-schema-ng: Implement bounding on types

### PR DESCRIPTION
**Pull request by jamesmunns**
This allows us to calculate max size.

This is a second attempt at what I started in #179.

### CODE REVIEW

File: source/postcard-schema-ng/src/key/hash.rs [hunk 1]
```diff
@@ -431,7 +469,7 @@ mod test {
 
         assert_eq!(
             hash_ty_path::<Bar>("test_path"),
-            [139, 128, 52, 27, 107, 8, 218, 98]
+            [224, 143, 54, 58, 255, 237, 252, 44]
```
Comment (jamesmunns) [inline source/postcard-schema-ng/src/key/hash.rs]: I don't know if I love this. I could leave `bounds: None` as not doing an additional hashing step, which would keep unbounded collections the same, but it would still be a breaking hash change for collections like `heapless::Vec<T, N>`.

File: source/postcard-schema-ng/src/impls/chrono_v0_4.rs [hunk 2]
```diff
@@ -4,5 +4,5 @@ use crate::{schema::DataModelType, Schema};
 
 #[cfg_attr(docsrs, doc(cfg(feature = "chrono-v0_4")))]
 impl<Tz: chrono_v0_4::TimeZone> Schema for chrono_v0_4::DateTime<Tz> {
-    const SCHEMA: &'static DataModelType = &DataModelType::String;
+    const SCHEMA: &'static DataModelType = &DataModelType::String { bounds: Some(32) };
```
Comment (max-heller) [inline source/postcard-schema-ng/src/impls/chrono_v0_4.rs]: How do we derive this bound? Would be helpful to add a comment

File: source/postcard-schema-ng/src/schema/mod.rs [hunk 3]
```diff
@@ -84,10 +81,16 @@ pub enum DataModelType {
     Char,
 
     /// The `String` Serde Data Model Type
-    String,
+    String {
+        /// Upper bound of items in the sequence
+        bounds: Option<usize>,
```
Comment (max-heller) [inline source/postcard-schema-ng/src/schema/mod.rs]: For all types: `bounds` to me implies multiple bounds (upper/lower); if we're only supporting an upper bound (which I think is reasonable, lower bounds do not seem nearly as useful except for things like calculating `min_size()` (https://github.com/jamesmunns/postcard/issues/180)), then I'd lean towards a name like `max_len` or `upper_bound` or something else that's more explicit.

For strings in particular: would be good to clarify that this is the number of bytes (similar to the docs of `str::len()`)

File: source/postcard-schema-ng/src/schema/mod.rs [hunk 4]
```diff
@@ -129,6 +139,120 @@ pub enum DataModelType {
...
+                    I128 => Some(postcard_core::varint::varint_max::<i128>()),
+                    U16 => Some(postcard_core::varint::varint_max::<u16>()),
+                    U32 => Some(postcard_core::varint::varint_max::<i32>()),
+                    U64 => Some(postcard_core::varint::varint_max::<u64>()),
+                    U128 => Some(postcard_core::varint::varint_max::<u128>()),
+                    Usize => Some(postcard_core::varint::varint_max::<usize>()),
+                    Isize => Some(postcard_core::varint::varint_max::<isize>()),
+                    F32 => Some(4),
+                    F64 => Some(8),
+                    Char => Some(5),
```
Comment (max-heller) [inline source/postcard-schema-ng/src/schema/mod.rs:187]: This comes from serializing a slice of up to four bytes, right? So one length byte plus up to four data bytes?

File: source/postcard-schema-ng/src/schema/mod.rs [hunk 5]
```diff
@@ -129,6 +139,120 @@ pub enum DataModelType {
...
+                        };
+                        let pair = ksize + vsize;
+                        let pairs = (*bound) * pair;
+                        Some(size_as_varint_usize(*bound) + pairs)
+                    }
+                    Struct { name: _, data } => data.max_size(),
+                    Enum { name: _, variants } => {
+                        let mut idx = 0;
+                        let disc = if variants.is_empty() {
+                            1
```
Comment (max-heller) [inline source/postcard-schema-ng/src/schema/mod.rs:232]: If there are no postcard-serializable variants, shouldn't it be impossible to serialize this enum, and therefore have a max size of 0? 1 seems like a fine fallback as well for this impossible case, just curious

File: source/postcard-schema-ng/src/schema/mod.rs [hunk 6]
```diff
@@ -129,6 +139,120 @@ pub enum DataModelType {
...
+                                return None;
+                            };
+                            if size > max {
+                                max = size;
+                            }
+                            idx += 1;
+                        }
+                        Some(disc + max)
+                    }
+                    Schema => todo!(),
```
Comment (max-heller) [inline source/postcard-schema-ng/src/schema/mod.rs]: Does `Schema` have a max size? Should this be `None`?

File: source/postcard-schema-ng/src/schema/mod.rs [hunk 7]
```diff
@@ -96,7 +99,12 @@ pub enum DataModelType {
     Unit,
 
     /// The "Sequence" Serde Data Model Type
-    Seq(&'static Self),
+    Seq {
+        /// Items in the sequence
+        item: &'static Self,
```
Comment (max-heller) [inline source/postcard-schema-ng/src/schema/mod.rs]: [`serde` uses `element` to refer to items in a sequence](https://docs.rs/serde/latest/serde/ser/trait.SerializeSeq.html#tymethod.serialize_element), would be nice to be consistent

File: source/postcard-schema-ng/src/schema/fmt.rs [hunk 8]
```diff
@@ -33,13 +33,16 @@ pub fn is_prim(osdmty: &OwnedDataModelType) -> bool {
         OwnedDataModelType::F32 => true,
         OwnedDataModelType::F64 => true,
         OwnedDataModelType::Char => true,
-        OwnedDataModelType::String => true,
-        OwnedDataModelType::ByteArray => true,
+        // TODO(AJM) is this right?
+        OwnedDataModelType::String { bounds } => bounds.is_none(),
+        OwnedDataModelType::ByteArray { bounds } => bounds.is_none(),
```
Comment (max-heller) [inline source/postcard-schema-ng/src/schema/fmt.rs]: What do we want `is_prim()` to guarantee? It's impossible for it to guarantee that the type whose schema we're looking at is literally `str` or `[u8]` (vs something like `String` or `Vec<u8>`), so I'd lean towards ignoring bounds here as `heapless::String<10>` feels comparable to `String`)

File: source/postcard-schema-ng/src/schema/fmt.rs [hunk 9]
```diff
@@ -105,18 +108,31 @@ pub fn fmt_owned_dmt_to_buf(dmt: &OwnedDataModelType, buf: &mut String, top_leve
         OwnedDataModelType::F32 => *buf += "f32",
         OwnedDataModelType::F64 => *buf += "f64",
         OwnedDataModelType::Char => *buf += "char",
-        OwnedDataModelType::String => *buf += "String",
-        OwnedDataModelType::ByteArray => *buf += "[u8]",
+        OwnedDataModelType::String { bounds } => {
+            *buf += "String";
```
Comment (max-heller) [inline source/postcard-schema-ng/src/schema/fmt.rs:114]: If we're doing a breaking release already, would be nice to change this to `str`

File: source/postcard-schema-ng/src/schema/fmt.rs [hunk 10]
```diff
@@ -105,18 +108,31 @@ pub fn fmt_owned_dmt_to_buf(dmt: &OwnedDataModelType, buf: &mut String, top_leve
...
+            if let Some(bound) = bounds {
+                *buf += &format!("[..{bound}]");
+            }
+        }
+        OwnedDataModelType::ByteArray { bounds } => {
+            *buf += "[u8]";
+            if let Some(bound) = bounds {
+                *buf += &format!("[..{bound}]");
```
Comment (max-heller) [inline source/postcard-schema-ng/src/schema/fmt.rs:122]: I don't have a strong preference either way, but we could consider something like `[u8; ..{bound}]` to be more concise. This would also work for sequences, but not for strings or maps, so maybe this uniform suffix is better

File: source/postcard-schema-ng/src/key/hash.rs (hunk 1 above)
Comment (max-heller) [reply source/postcard-schema-ng/src/key/hash.rs]: My preference here would be to hash only the pieces that break wire compatibility if changed (i.e. only the serde-relevant pieces), at least by default

Prior discussion: https://github.com/jamesmunns/postcard/issues/192#issuecomment-2506726286

**Comment by max-heller** [comment]
Thanks for reviving this! Very excited to remove our `MaxSize` derive

File: source/postcard-schema-ng/src/impls/chrono_v0_4.rs (hunk 2 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/impls/chrono_v0_4.rs]: Ah, I should. I looked at chrono's rfc3339 output, and counted the output in the worst case (with nanos precision), which IIRC was 29 or 30, and I added a fudge factor. I should actually add a test case here for this though and not round up.

File: source/postcard-schema-ng/src/schema/mod.rs (hunk 3 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/schema/mod.rs]: I think that's reasonable, it is definitely just an "upper bounds".

I feel like "MinSize" would be basically the same as the `fn max_size` impl with all of the current upper bounds treated as zero?.

File: source/postcard-schema-ng/src/schema/mod.rs (hunk 4 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/schema/mod.rs:187]: Yep, first byte can be 1..=4, then the 1..=4 bytes of the char. (this is a terrible encoding form, there's an issue to fix this if we ever do a breaking wire change).

File: source/postcard-schema-ng/src/schema/mod.rs (hunk 5 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/schema/mod.rs:232]: That's a good point. I might try to add a test around this, but an enum with zero variants does seem a bit silly, wire wise.

File: source/postcard-schema-ng/src/schema/mod.rs (hunk 6 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/schema/mod.rs]: This should probably be `None`, yeah.

File: source/postcard-schema-ng/src/schema/mod.rs (hunk 3 above)
Comment (max-heller) [reply source/postcard-schema-ng/src/schema/mod.rs]: Leaning towards `max_len` over `upper_bound` so the same name makes sense for strings as well as other collections

Comment (max-heller) [reply source/postcard-schema-ng/src/schema/mod.rs]: > I feel like "MinSize" would be basically the same as the fn max_size impl with all of the current upper bounds treated as zero?.

It can be, though in the presence of types like `NonEmptyVec` or similar it would be an underestimate

File: source/postcard-schema-ng/src/schema/fmt.rs (hunk 8 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/schema/fmt.rs]: I need to look again what I was using this for. I think I used it for postcard-rpc or poststation to avoid sending "builtin" schemas, this may be different now in postcard-schema-ng where the shape of schemas have changed and we don't have top level types like `u8` containing a named data type.

I may just remove it for now, we can add it back later if we need.

File: source/postcard-schema-ng/src/schema/fmt.rs (hunk 10 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/schema/fmt.rs:122]: oooh, that's a better syntax, and yeah, I struggled with what this should be.

File: source/postcard-schema-ng/src/key/hash.rs (hunk 1 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/key/hash.rs]: Yeah, let me read that back. At least from the perspective of postcard-rpc, I am somewhat tempted to say "if you change the max size, that's a wire compatibility change".

I know this is sort of a grey area though, if you HAPPEN to have enough space it totally will decode right (and therefore "the wire representations are the same"), but from the perspective of "do both sides agree on the same schema", I think different sizes should be treated as different hashes.

Comment (max-heller) [reply source/postcard-schema-ng/src/key/hash.rs]: At $WORK we depend on being able to change the max length in a backwards-compatible way in some places (we have a dedicated `TruncatingVec` type that will accept as many elements as it can fit). Going back to the '24 discussion, we could expose options for how to calculate the hash, and allow users to decide what they want? If postcard-rpc wants to take the strict interpretation that seems like a sane answer, but I don't think we should force it on all users

Comment (jamesmunns) [reply source/postcard-schema-ng/src/key/hash.rs]: Yeah, I think the right way to handle this is to change the hasher to include a `config` item (in addition to the current u64 state), and allow the caller to pass in that config which decides whether bounds (or future metadata) is considered as part of the hashing. I haven't done that yet, but will.

File: source/postcard-schema-ng/src/impls/chrono_v0_4.rs (hunk 2 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/impls/chrono_v0_4.rs]: Added a test for this, 32 was underconsidering it. I think I have an upper bound now, but it's potentially questionable whether we consider this bounded or not.

File: source/postcard-schema-ng/src/schema/mod.rs (hunk 3 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/schema/mod.rs]: Renamed to max_len, and added `MaxLen*` types that include these bounds.

Good call on the NonEmpty items. Maybe some day we want to add `min_len` as metadata, I don't totally understand the desire for it though practically, so maybe I'll leave it for a future PR. I wonder if we should make some more things `non_exhaustive` to avoid this being a breaking change.

File: source/postcard-schema-ng/src/schema/mod.rs (hunk 6 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/schema/mod.rs]: Fixed.

File: source/postcard-schema-ng/src/schema/fmt.rs (hunk 9 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/schema/fmt.rs:114]: By the way, why "str" and not "String"? I feel like either are equally valid.

**Comment by jamesmunns** [comment]
I do still need to update postcard-dyn.

File: source/postcard-schema-ng/src/key/hash.rs (hunk 1 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/key/hash.rs]: @max-heller one thing that I realized when thinking about this: if you're doing what you describe, and you start using the `max_size()` stuff to auto-size buffers, won't that end up biting you if someone sends an array that's too big, and that pushes the size of a single message over the calculated `max_size()`?

This may not be a problem if you're using something else as the max length, like if you say a frame can't exceed an MTU of 1544 bytes or something, which ends up being the forcing function instead. But if you you're buffering serial data for example, and with a `TruncatingVec<24>` you calculate your max RX size as 300 bytes, but someone sends a `TruncatingVec<32>` and ends up sending you a 400 byte message, you'll overflow your incoming buffer entirely and still discard the message.

I still think "add a config to the hasher" is probably the right take, but it does seem *somewhat* foot-gunny to use auto-sized buffers using `max_size` *without* enabling the hashing of bounds.

File: source/postcard-schema-ng/src/impls/uuid_v1_0.rs [hunk 11]
```diff
@@ -3,5 +3,20 @@
 use crate::{schema::DataModelType, Schema};
 
 impl Schema for uuid_v1_0::Uuid {
-    const SCHEMA: &'static DataModelType = &DataModelType::ByteArray { bounds: Some(16) };
+    const SCHEMA: &'static DataModelType = &DataModelType::Seq {
```
Comment (max-heller) [inline source/postcard-schema-ng/src/impls/uuid_v1_0.rs:6]: I think `ByteArray` was correct given that the implementation uses `serialize_bytes()`: https://docs.rs/uuid/1.26.0/src/uuid/external/serde_support.rs.html#31

File: source/postcard-schema-ng/src/schema/fmt.rs (hunk 9 above)
Comment (max-heller) [reply source/postcard-schema-ng/src/schema/fmt.rs:114]: `str` is to `String` what `[u8]` (which we use for `ByteArray`) is to `Vec<u8>` (or `Box<[u8]>`), which is to say the underlying primitive type -- and serde data model type

File: source/postcard-schema-ng/src/impls/chrono_v0_4.rs (hunk 2 above)
Comment (max-heller) [reply source/postcard-schema-ng/src/impls/chrono_v0_4.rs]: I think it's fine either way. I'd lean towards avoiding the bound since RFC3339 seems to technically allow unbounded length

File: source/postcard-schema-ng/src/schema/mod.rs (hunk 3 above)
Comment (max-heller) [reply source/postcard-schema-ng/src/schema/mod.rs]: > so maybe I'll leave it for a future PR

Agreed this makes sense for a separate PR if at all

> I wonder if we should make some more things non_exhaustive to avoid this being a breaking change.

This could make sense, though we should avoid making it a pain to manually implement `Schema`. One option would be to take inspiration from [`syn` 3.0](https://github.com/dtolnay/syn/releases/tag/3.0.0) (see "Modifiers") and add `non_exhaustive` structs as fields in places we expect to want to add more data in the future, see e.g. https://docs.rs/syn/3.0.5/syn/struct.Field.html#structfield.modifiers

File: source/postcard-schema-ng/src/key/hash.rs [hunk 12]
```diff
@@ -65,6 +65,24 @@ pub mod fnv1a64 {
         hash_sdm_type(state, schema).to_le_bytes()
     }
 
+    pub(crate) const fn hash_bounds(state: u64, bounds: Option<usize>) -> u64 {
+        let Some(bound) = bounds else {
+            return state;
+        };
+        let mut state = hash_update(state, &[0x2B]);
```
Comment (max-heller) [inline source/postcard-schema-ng/src/key/hash.rs:72]: When we are going to be hashing bounds (according to the hashing options), don't we need to hash `0x2B` everywhere the optional bounds appear (even if `None`) to avoid hash collisions? If we only hash this prime when bounds *are* present, I think we can get into ambiguities where something like `{ name: [0x1], max_len: Some(0x5) }` feeds the same bytes into the hasher as `{ name: [0x1, 0x2B, 0x5], max_len: None }`

File: source/postcard-schema-ng/src/key/hash.rs (hunk 1 above)
Comment (max-heller) [reply source/postcard-schema-ng/src/key/hash.rs]: That's a good point, that can certainly bite you. On the one hand, that outcome is bad, so it's worth discouraging; on the other, if we require hashing the max length, then the message would have been rejected anyway because of a hash mismatch, so you aren't much better off.

In practice, I think for things like this we usually have a fixed buffer size (MTU/max IP packet size/etc.) or a max size that far exceeds the expected size of any given packet because of many `None`s inside.

The other use-case that I find valuable that depends on ignoring max length in hashing is sending `&[T]` where `heapless::Vec<T, N>` is expected (and other equivalent pairs). This avoids needing to stack-allocate a `Vec` just to serialize it again. This could probably be solved via bounded, borrowed slice types like the ones you've added, but that may not work in every case.

File: source/postcard-schema-ng/src/impls/uuid_v1_0.rs (hunk 11 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/impls/uuid_v1_0.rs:6]: Yeah... postcard doesn't actually discriminate well between "byte array" and "[u8]", and we declare the schema of `[u8]` as

```
impl<T: Schema> Schema for [T] {
    const SCHEMA: &'static DataModelType = &DataModelType::Seq {
        element: T::SCHEMA,
        max_len: None,
    };
}
```

I wonder if we should actually just discard `ByteArray` in `-ng` to avoid this being confusing.

File: source/postcard-schema-ng/src/key/hash.rs (hunk 12 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/key/hash.rs:72]: I think https://github.com/jamesmunns/postcard/issues/217 captures this, we should probably fix that for -ng.

File: source/postcard-schema-ng/src/impls/uuid_v1_0.rs (hunk 11 above)
Comment (max-heller) [reply source/postcard-schema-ng/src/impls/uuid_v1_0.rs:6]: Given [it's a distinct type in the data model](https://serde.rs/data-model.html#types), I think it's important to keep `ByteArray` around -- serializers can choose to serialize them differently than a sequence of bytes (though Postcard doesn't)

Comment (max-heller) [reply source/postcard-schema-ng/src/impls/uuid_v1_0.rs:6]: The fact that `[u8]` can't be specialized to use `serialize_bytes()` is more of an implementation limitation than anything

File: source/postcard-schema-ng/src/key/hash.rs (hunk 12 above)
Comment (max-heller) [reply source/postcard-schema-ng/src/key/hash.rs:72]: That probably covers it... though it still gives me a bad feeling to do it conditionally

File: source/postcard-schema-ng/src/impls/uuid_v1_0.rs (hunk 11 above)
Comment (jamesmunns) [reply source/postcard-schema-ng/src/impls/uuid_v1_0.rs:6]: > serializers can choose to serialize them differently than a sequence of bytes (though Postcard doesn't)

Are you using `postcard-schema` for not-postcard-format things? A long while back you asked "is postcard-schema for postcard only, or serde in general", and these days, I'm feeling a bit stronger about it only being for postcard.

Comment (max-heller) [reply source/postcard-schema-ng/src/impls/uuid_v1_0.rs]: > Are you using postcard-schema for not-postcard-format things?

Kinda sorta not really. `postcard-schema` is *close* to being useful as a general `serde` schema, but I've somewhat lost hope in full fledged dynamic (de)serialization with serde. Where possible though (including here), I think it's still valuable to have a 1:1 match between `postcard-schema` and `serde`'s data model or other abstractions.
